---
title: "Simulating plot data in a plant breeding field trial - FieldSimR demonstration"
author: "D.J. Tolhurst and C.R. Werner"
output: html_document
vignette: >
  %\VignetteIndexEntry{Simulating plot data in a plant breeding field trial - FieldSimR demonstration}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

<br/>

The objective of this tutorial is to simulate plot data using FieldSimR which captures spatial variation with realistic structure and complexity. Genetic values will be simulated using AlphaSimR based on an additive genetic trait representing grain yield.

<br/>

<div class="warning" style='background-colour: rgba(0,0,0,0.07); colour: #000000; border-left: solid #000080 4px; border-right: solid #000080 4px; border-top: solid #000080 4px; border-bottom: solid #000080 4px; border-radius: 7px; padding:0.7em;'>
<span>
<p style='margin-top:0.5em; margin-left:0.5em; text-align:left; font-size:25px'>
**Summary** </b></p>
<p style='margin-left:1em;'>
The simulation can be summarised by the following steps:
</p>
1.  Simulating plot errors with FieldSimR
    a. Simulation parameters
    b. `field_trial_error()` and `plot_effects()` functions
  
</p>
2.  Simulating genetic values with AlphaSimR
    a. Simulation parameters
    b. Simulating an additive genetic trait
    
</p>
3.  Generating phenotypes by combining the plot errors with the genetic values
    a.  `make_phenotypes()` function
        - Measures of heritability and expected accuracy
    
</p>
<p style='margin-bottom:1.5em;'>
</p></span>
</div>

<br/>

### **Preliminaries**

Clean the working environment.

```{r}
rm(list = ls())
```

Install FieldSimR from github.

```{r, eval = FALSE}
library(devtools)
install_github("crWerner/fieldsimr")
```

Load required packages.

```{r, results = 'hide'}
library(FieldSimR)
library(AlphaSimR)
library(ggplot2)
```

The results in this tutorial can be reproduced by initially setting the following seed.

```{r}
set.seed(123)
```

<br/>

### **Simulation details**

Consider a scenario in which 100 inbred wheat genotypes are evaluated for grain yield (t/ha) in a single field trial. The genotypes are simulated with additive genetic relatedness using AlphaSimR. The trial comprises 10 columns and 20 rows for 200 plots in total. There are two complete blocks of all genotypes aligned in the column direction (side-by-side). The trial is 80 m long in the column direction and 40 m wide in the row direction, with plots 8m long by 2m wide. The trait mean is 4 t/ha and the trait variance is 0.2 t<sup>2</sup>/ha<sup>2</sup>. The plot-level heritability is 0.3, which reflects the ratio of genetic variance to total phenotypic variance. The error variance is given by 0.2/0.3-0.2 = 0.47 t<sup>2</sup>/ha<sup>2</sup>.

<br/> 

The simulation parameters are:

```{r}
ngenos <- 100 # no. genotypes
nreps <- nblocks <- 2 # no. replicates/blocks
block_dir <- "col" # block direction
ncols <- 10   # no. columns
nrows <- 20   # no. rows
nplots <- ncols * nrows # no. plots == ngenos * nreps
plot_length <- 8
plot_width <- 2 
mu <- 4        # trait mean
sigma2 <- 0.2  # trait variance 
H2 <- 0.3      # plot-level heritability
sigma2e <- sigma2/H2 - sigma2 # error variance
round(sigma2e, digits = 2)
```

<br/> 

The blank field trial layout is:

```{r, echo = FALSE, fig.width = 9.5, fig.height = 4.5}
layout_df <- data.frame(block = factor(rep(1:2, each = nplots/2)),
                        col = factor(rep(1:ncols, each = nrows)),
                        row = factor(1:nrows),
                        plot = factor(1:nplots))
FieldSimR::plot_effects(layout_df, effect = "block")
```

<br/>

The phenotypes are generated based on the linear mixed model:

$\mathbf{y} = \mathbf{Z}\mathbf{gv} + \mathbf{e}$,

where $\mathbf{gv}$ is the vector of genetic values with design matrix $\mathbf{Z}$ 
and $\mathbf{e}$ is the vector of plot errors. The inclusion of additional non-genetic effects is straightforward.

The simulation involves three steps:

1.  Simulating the plot errors with FieldSimR
2.  Simulating the genetic values with AlphaSimR
3.  Generating the phenotypes by combining the plot errors with the genetic values.

<br/>

### **1. Simulating the plot errors with FieldSimR**

The plot errors are simulated to capture spatial variation, which can be broadly categorised as either global and local trend (smooth spatial trend), random error (noise) or extraneous variation (systematic error).

The plot errors are constructed as the sum of three terms:

$\mathbf{e} = \mathbf{se} + \mathbf{re} + \mathbf{ee}$,

where $\mathbf{se}$ is a vector of _spatial_ errors that capture global and local trend,
$\mathbf{re}$ is a vector of _random_ errors that capture random measurement error/noise
and $\mathbf{ee}$ is a vector of _extraneous_ errors that capture extraneous variation.

Simulating the plot errors involves two parts:

a. Setting the simulation parameters for the spatial, extraneous and random error terms
b. Simulating the plot errors with the `field_trial_error()` function and displaying them with the `plot_effects()` function. 

<br/>

#### **a. Setting the simulation parameters**

The plot errors are generated based on a proportion of spatial error variance of 0.4 and a proportion of extraneous error variance of 0.2. FieldSimR will then allocate the remaining 0.4 to the random error variance by default.

The initial simulation parameters are:

```{r}
round(sigma2e, digits = 2) # error variance
prop_spatial <- 0.4 # proportion of spatial error variance
prop_ext <- 0.2 # proportion of extraneous error variance
1 - (prop_spatial + prop_ext) # proportion of random error variance
```

The simulation parameters for the spatial and extraneous errors are defined in the following. No further parameters are required for the random errors.

<br/>

##### **Spatial errors**

FieldSimR has the capacity to simulate spatial errors based on a separable first order autoregressive (AR1) process or  bivariate interpolation. These options are set with `spatial.model = "AR1"` or `spatial.model = "Bivariate"`.

FieldSimR's functionality will be demonstrated using a separable AR1 process with column autocorrelation of 0.5 and row autocorrelation of 0.7. The simulation parameters are:

```{r}
spatial_model1 <- "AR1"
prop_spatial # proportion of spatial error variance
col_cor <- 0.5 # column autocorrelation
row_cor <- 0.7 # row autocorrelation
```
Note that the plot lengths and widths are not required when `spatial.model = "AR1"`.

<br/>

FieldSimR's functionality will also be demonstrated using bivariate interpolation with 100 additional knots points. This option is set with `complexity = 100`. The simulation parameters are:

```{r}
spatial_model2 <- "Bivariate"
prop_spatial # proportion of spatial error variance
complexity <- 100 # number of additional knot points
plot_length
plot_width
```
Note that increasing the complexity will generate more local trend relative to global trend, up to a point where the spatial errors capture minimal or no trend.

<br/>

##### **Extraneous errors**

FieldSimR has the capacity to simulate extraneous errors based on random or zig-zag ordering. These options are set with `ext.ord = "random"` or `ext.ord = "zig-zag"`. The zig-zag ordering is achieved by alternating the positive and negative simulated values between neighbouring columns and/or rows.

Extraneous errors can be simulated in both directions, or either the column or row directions individually.
These options are set with `ext.dir = "both"`, `ext.dir = "col"` or `ext.dir = "row"`.

The functionality will be demonstrated using zig-zag ordering between neighbouring rows. The simulation parameters are:

```{r}
prop_ext # proportion of extraneous error variance
ext_ord <- "zig-zag" # ordering of extraneous variation
ext_dir <- "row" # direction of extraneous variation
```

<br/>

#### **b. Simulating the plot errors**

The plot errors are simulated using FieldSimR's function `field_trial_error()`. The functionality will be demonstrated separately for the simulations involving the AR1 process and bivariate interpolation.

<br/>

##### **Autoregressive process**

The simulation is achieved by populating the `field_trial_error()` function with terms appropriate to the AR1 process, i.e. `col.cor` and `row.cor`.
A data frame is returned containing the environment, block, column and row numbers as well as the simulated plot error. 
When `return.effects = TRUE`, a list is returned with an additional data frame containing the spatial, random and extraneous error terms.

```{r}
error_df1 <- field_trial_error(ntraits = 1,
                               nenvs = 1,
                               nblocks = nblocks, # 2
                               block.dir = block_dir, # "col"
                               ncols = ncols, # 10
                               nrows = nrows, # 20
                               varR = sigma2e, # 0.47
                               spatial.model = spatial_model1, # "AR1"
                               col.cor = col_cor, # 0.5
                               row.cor = row_cor, # 0.7
                               prop.spatial = prop_spatial, # 0.4
                               ext.ord = ext_ord, # "zig-zag"
                               ext.dir = ext_dir, # "row"
                               prop.ext = prop_ext, # 0.2
                               return.effects = TRUE)
head(error_df1$error.df) # data frame with simulated plot errors
head(error_df1$Trait1) # data frame with simulated spatial, random and extraneous errors
```

<br/>

The separate error terms can be displayed using FieldSimR's function `plot_effects()`.
For example, the spatial errors are given by:

```{r, fig.width = 9.5, fig.height = 4.5}
plot_effects(error_df1$Trait1, effect = "e.spat", labels = TRUE)
```

<br/>

The total plot errors are given by:

```{r, fig.width = 9.5, fig.height = 4.5}
plot_effects(error_df1$error.df, effect = "e.Trait1", labels = TRUE)
```

<br/>

The theoretical and sample variograms can be displayed using FieldSimR’s functions `theoretical_variogram()` and `sample_variogram()`. 
The sample variogram for the total plot errors is given by:

```{r, fig.width = 9.5, fig.height = 4.5}
sample_variogram(error_df1$error.df, effect = "e.Trait1")
```

##### **Bivariate interpolation**

The simulation is achieved by populating the `field_trial_error()` function with terms appropriate to bivariate interpolation, i.e. `plot.length`, `plot.width` and `complexity`.

```{r}
error_df2 <- field_trial_error(ntraits = 1,
                               nenvs = 1,
                               nblocks = nblocks, # 2
                               block.dir = block_dir, # "col"
                               ncols = ncols, # 10
                               nrows = nrows, # 20
                               varR = sigma2e, # 0.47
                               prop.spatial = prop_spatial, # 0.4
                               spatial.model = spatial_model2, # "Bivariate"
                               plot.length = plot_length, # 8
                               plot.width = plot_width, # 2
                               complexity = complexity, # 100
                               prop.ext = prop_ext, # 0.2
                               ext.dir = ext_dir, # "row"
                               ext.ord = ext_ord, # "zig-zag"
                               return.effects = TRUE)
head(error_df2$error.df) # data frame with simulated plot errors
head(error_df2$Trait1) # data frame with simulated spatial, random and extraneous errors
```

<br/>

The spatial errors are given by:

```{r, fig.width = 9.5, fig.height = 4.5}
plot_effects(error_df2$Trait1, effect = "e.spat", labels = TRUE)
```

<br/>

The total plot errors are given by:

```{r, fig.width = 9.5, fig.height = 4.5}
plot_effects(error_df2$error.df, effect = "e.Trait1", labels = TRUE)
```

<br/>

### **2. Simulating the genetic values with AlphaSimR**

Simulating the genetic values involves two parts:

a. Setting the simulation parameters for the founder population and trait architecture
b. Simulating an additive genetic trait for a population of inbred wheat genotypes. 

<br/>

#### **a. Setting the simulation parameters**

The number of homogeneous genotypes in the founder population is set to 20, the number of chromosomes is set to 21 and the number of segregating sites (biallelic QTN) per chromosome is set to 300. 

```{r}
nfounders <- 20 # Number of genotypes in the founder population
nchr <- 21 # Number of chromosomes
nseg_sites <- 300 # Number of QTN per chromosome
```

<br/>

The trait mean and variance are:

```{r}
mu # trait mean
sigma2 # trait variance
```

<br/>

#### **b. Simulating an additive genetic trait**

The founder population is simulated using the `runMacs()` function.

```{r}
# Simulating a founder population using AlphaSimR's "WHEAT" presets to mimic it's evolutionary history
founders <- runMacs(nInd = nfounders,
                    nChr = nchr,
                    segSites = nseg_sites,
                    inbred = TRUE,
                    species = "WHEAT",
                    nThreads = 2)
founders # MapPop object
SP <- SimParam$new(founders)
```

<br/>

An additive genetic trait representing grain yield is then simulated using the `addTraitA()` function.

```{r}
SP$addTraitA(nQtlPerChr = nseg_sites,
             mean = mu,
             var = sigma2)
founders <- newPop(founders) # create pop-class object
```

<br/>

The inbred wheat genotypes are generated by randomly crossing the founders using the `randCross()` function, and then made into doubled haploid (DH) lines using the `makeDH()` function.

```{r}
f1s <- randCross(pop = founders, 
                 nCrosses = 20, # no. crosses
                 nProgeny = 5) # no. progeny per cross
DHs <- makeDH(pop = f1s, nDH = 1)
DHs
```

<br/>

The genetic values are obtained for the DH population.

```{r}
gv <- c(DHs@gv)
round(mean(gv), digits = 2) # trait mean in the DH population
round(var(gv), digits = 2) # trait variance in the DH population
gv_df <- data.frame(env = factor(1),
                    rep = factor(rep(1:nreps, each = ngenos)),
                    id = factor(DHs@id),
                    gv = gv)
head(gv_df)
```

<br/>

### **3. Generating the phenotypes**

The phenotypes are generated using FieldSimR's function `make_phenotypes()`. 
When `randomise = TRUE`, genotypes are allocated to plots using a randomised complete block design.

```{r}
trial_ls <- make_phenotypes(gv.df = gv_df, error.df = error_df2, randomise = TRUE, return.effects = TRUE)
trial_df <- trial_ls$pheno.df
head(trial_df)
```

The true genetic values are also added to aid with interpretation.

```{r}
trial_df$gv.Trait1 <- trial_ls$Trait1$gv.Trait1
head(trial_df)
```

<br/>

The phenotypes are given by:

```{r, fig.width = 9.5, fig.height = 4.5}
plot_effects(trial_df, effect = "y.Trait1", labels = TRUE)
```

<br/>

##### **c. Measures of heritability and expected accuracy**

The simulated plot-level and line-mean heritabilities can be formally quantified to aid with interpretation. The square-root of the latter represents the expected prediction accuracy for the genetic values.

A quick look at the phenotypes and true genetic values.

```{r, echo = FALSE}
head(trial_df)
ggplot(trial_df, aes(x = gv.Trait1, y = y.Trait1)) +
    geom_hline(yintercept = mu, linetype = 2) + geom_vline(xintercept = mu, linetype = 2) + geom_point() +
    labs(x = "Genetic values for grain yield (t/ha)", y = "Phenotypes for grain yield (t/ha)")
```

The plot-level heritability is:

```{r}
round(cor(trial_df$y.Trait1, trial_df$gv.Trait1)^2, digits = 2) # simulated plot-level heritability
round(sigma2/(sigma2 + sigma2e), digits = 2) # expected plot-level heritability
```
The minor discrepancies here are a result of the sampling process.

<br/>

A quick look at the average phenotypes and true genetic values.

```{r}
index <- trimws(trial_df$id[trial_df$block == 1]) # obtain genotype ordering
ave_yield <- with(trial_df, tapply(y.Trait1, id, mean))[index]
```

```{r, echo = FALSE}
ggplot() + geom_hline(yintercept = mu, linetype = 2) + geom_vline(xintercept = mu, linetype = 2) + 
    geom_point(data = trial_df, aes(x = gv.Trait1, y = y.Trait1), alpha = 0.1) + 
    geom_point(aes(x = trial_df$gv.Trait1[is.element(trial_df$block, 1)], y = ave_yield)) +
    labs(x = "Genetic values for grain yield (t/ha)", y = "Phenotypes for grain yield (t/ha)")
```

The line-mean heritability in the DH population is:

```{r}
round(cor(ave_yield, trial_df$gv.Trait1[is.element(trial_df$block, 1)])^2, digits = 2) # simulated line-mean heritability
round(sigma2/(sigma2 + sigma2e/nreps), digits = 2) # expected line-mean heritability
round(cor(ave_yield, trial_df$gv.Trait1[is.element(trial_df$block, 1)]), digits = 2) # simulated prediction accuracy
round(sqrt(sigma2/(sigma2 + sigma2e/nreps)), digits = 2) # expected prediction accuracy
```

<br/>

<br/>



